airflow/operators/singleton_operator.py

```diff
@@ -0,0 +1,109 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+#     specific language governing permissions and limitations
+# under the License.
+
+from airflow.models import BaseOperator, DagRun, TaskInstance
+from airflow.models.skipmixin import SkipMixin
+from airflow.utils.db import provide_session
+from airflow.utils.decorators import apply_defaults
+from airflow.utils.state import State
+
+
+class SingletonOperator(BaseOperator, SkipMixin):
+    """
+    Allows a workflow to ensure that only single `unfinished` task is being
+    executed at all times while allowing airflow to schedule multiple dags to
+    execute other task instances that are not directly downstream of the task.
+
+    Unfinished task states:
+            NONE,
+            SCHEDULED,
+            QUEUED,
+            RUNNING,
+            UP_FOR_RETRY
+
+    All directly downstream tasks will be skipped.
+
+    :param allow_external_trigger: If this is set, execution is allowed
+        regardless of dag_run states.
+    :type allow_external_trigger: Boolean
+    """
+
+    ui_color = '#e9ffdb'  # nyanza
+
+    @apply_defaults
+    def __init__(self, allow_external_trigger=False, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.task_instance = None
+        self.dag_run = None
+        self.downstream_task_instances = None
+        self.downstream_task_instance_ids = None
+        self.allow_external_trigger = allow_external_trigger
+
+    def execute(self, context):
+        if self.allow_external_trigger and context['dag_run'].external_trigger:
+            self.log.info("Externally triggered dag_run: allowing execution.")
+            return State.SCHEDULED.upper()
+
+        self.task_instance = context['ti']
+        self.dag_run = context['dag_run']
+        self.downstream_task_instances = self.get_flat_relatives(upstream=False)
+        self.downstream_task_instance_ids = [ti.task_id for ti in
```

Comment:
The BaseOperator has methods (properties) `downstream_task_ids()` and `upstream_task_ids()` to get the same, could you use those?

Reply:
Absolutely. I will update it.